Make return type of set_l*_indices be a renamedt [blocks: #4333]

romainbrenguier opened the pull request:
This makes it easier to track through which kind of renaming the symbols have been through.
This is an intermediary step, for getting all renamed expression use this types (the next step would be to do the same for all the `rename_*` functions)

Edit: I have changed the title of the PR to better reflect its content now that it has been rebased on the preliminary PR introducing the renamedt type

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [na] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

martin-cs reviewed (commented):
This seems like quite a big step.  How does this interact with the validity checks for programs, expressions, etc. that is being developed to capture those system invariants?

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -237,11 +237,13 @@ void goto_symex_statet::assignment(
 
-void goto_symex_statet::set_l0_indices(
+static void set_l0_indices(
+  symex_level0t level0,
```
Should probably be by-ref?

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -337,9 +336,20 @@ void goto_symex_statet::rename_level0(exprt &expr, const namespacet &ns)
 }
 
+ssa_exprt
+goto_symex_statet::rename_level0(const symbol_exprt &expr, const namespacet &ns)
```
Shouldn't that be specifically an `ssa_exprt`, considering we still explicitly convert symbol_exprt -> ssa_exprt in `rename_*`?

smowton commented:
I think the template argument for `level0t` et al would always have to be `ssa_exprt`, so I'd recommend just making these subclasses of `exprt`. Note I would keep the type-incompatibility of `l0_ssa_exprt` and `l1_ssa_exprt`, since it's important that an L1 expr can't be silently cast to an L0 expr.

smowton commented:
@martin-cs these would presumably render such invariants unnecessary when we can enforce with the type system, similar to how we would pass a `struct_typet` around and not constantly re-verify that it meets the required structure of such a type (that was checked by `to_struct_typet`)

romainbrenguier commented:
@smowton 
> I think the template argument for level0t et al would always have to be ssa_exprt, so I'd recommend just making these subclasses of exprt. Note I would keep the type-incompatibility of l0_ssa_exprt and l1_ssa_exprt, since it's important that an L1 expr can't be silently cast to an L0 expr.

The argument sometimes needs to be exprt because for instance we have `if_exprt` or others which have been renamed. On the other hand ssa_exprt can only represent symbols.

romainbrenguier committed:
Make set_l0_indices return a renamedt

This makes explicit that the result is a renamed expression and can be
propagated in through other functions.

romainbrenguier committed:
Make set_l1_indices return a renamedt

This carries the information about which renaming the expression has
been through.
This requires a change in symex_level1t::operator() so that it takes
care of the non-empty level 2 case.

romainbrenguier committed:
Make set_l2_indices return a renamedt

This captures the information of which renaming the expression has been
through.

This requires making symex_level2t::operator() handle non-empty L2. This
was handled by set_l2_indices, but it is neccessary to handle it in
symex_level2t for making set_l2_indices return a renamedt.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: f13a9461d2ee32e761284b2e67778f0273697431).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103232156
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -188,19 +188,22 @@ void goto_symex_statet::assignment(
 
   // in case we happen to be multi-threaded, record the memory access
-  bool is_shared=l2_thread_write_encoding(lhs, ns);
+  bool is_shared = l2_thread_write_encoding(l2_lhs.get(), ns);
```
It seems `lhs == l2_lhs.get()`, so why can't you just keep using `lhs` here (and below)?

tautschnig commented on src/goto-symex/renaming_level.cpp:
```diff
@@ -46,8 +46,14 @@ operator()(ssa_exprt ssa_expr, const namespacet &ns, unsigned thread_nr) const
+    !l0_expr.get().get_level_2().empty())
+  {
+    // TODO reaching here means the information that l0_expr was L1 or L2 has
+    // TODO been lost somewhere during execution, which should be avoided
```
Use `UNREACHABLE` (here and below)?

romainbrenguier replied:
Unfortunately it is reachable. There are even cases where level 2 is set but not level 1.

tautschnig replied:
Level 2 being set without level 1 is expected for objects with static lifetime, but maybe there are also actually buggy cases?

romainbrenguier replied:
That's good to know. I think it will be easier to see if something is done wrong once the renamedt type is used everywhere in goto_symex_state. (just to be clear, we do not change the behaviour here, the ignoring of the non-empty case was just done in the caller of this function).

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: d723b64a091e00313ec4740e2cc4a098fec1b5f5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103247786
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

tautschnig commented:
1) I think these commits either need squashing or forcibly updated dates - reviewing commit-by-commit on GitHub is not meaningful at the moment, and I'm also finding it very hard to figure out from the commit subjects where I might have to start.
2) I don't think that a seemingly nicely working code should newly be littered with TODOs. That's not an improvement. If there is a problem, then it should be debugged and fixed.

romainbrenguier commented:
@tautschnig I squashed three commits, the order looks fine for me now. I removed the TODOs since it seems there is nothing wrong there.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -275,7 +272,10 @@ ssa_exprt goto_symex_statet::rename_ssa(ssa_exprt ssa, const namespacet &ns)
-    set_l0_indices(ssa, ns);
+  {
+    const renamedt<ssa_exprt, L0> ssa_l0 = set_l0_indices(std::move(ssa), ns);
+    ssa = ssa_l0.get();
```
Does the temporary really add value? I think it's pretty clear what `set_l0_indices` does. Therefore I'd prefer `ssa = set_l0_indices(std::move(ssa), ns).get();`

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -277,7 +270,10 @@ ssa_exprt goto_symex_statet::rename_ssa(ssa_exprt ssa, const namespacet &ns)
-    set_l1_indices(ssa, ns);
+  {
+    const renamedt<ssa_exprt, L1> ssa_l1 = set_l1_indices(std::move(ssa), ns);
+    ssa = ssa_l1.get();
```
Like for the level-0 case, I'd prefer `ssa = set_l1_indices(std::move(ssa), ns).get();`

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -312,7 +308,8 @@ exprt goto_symex_statet::rename(exprt expr, const namespacet &ns)
     {
-      set_l1_indices(ssa, ns);
+      const renamedt<ssa_exprt, L1> ssa_l1 = set_l1_indices(std::move(ssa), ns);
+      ssa = ssa_l1.get();
```
As above, `ssa_l1` is not necessary.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -563,8 +560,8 @@ void goto_symex_statet::rename_address(exprt &expr, const namespacet &ns)
-    set_l1_indices(ssa, ns);
-
+    const renamedt<ssa_exprt, L1> ssa_l1 = set_l1_indices(std::move(ssa), ns);
+    ssa = ssa_l1.get();
```
As above.

tautschnig commented on src/goto-symex/renaming_level.cpp:
```diff
@@ -46,8 +46,12 @@ operator()(ssa_exprt ssa_expr, const namespacet &ns, unsigned thread_nr) const
-  if(!l0_expr.get().get_level_1().empty())
+  if(
+    !l0_expr.get().get_level_1().empty() ||
+    !l0_expr.get().get_level_2().empty())
```
To make sure there is no change in behaviour this will require an `INVARIANT` that `!l0_expr.get().get_level_2().empty()` implies `!l0_expr.get().get_level_1().empty()`. And I'm not convinced that holds, because a global variable would not need L1 renaming, even when L2 is non-empty.

romainbrenguier replied:
Yes this does not hold, so we can't add an invariant. The reason this does not change the behaviour is that it is only called from `set_l1_indices` which had the same check until this commit.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -188,7 +188,8 @@ void goto_symex_statet::assignment(
   symex_renaming_levelt::increase_counter(level2_it);
-  set_l2_indices(lhs, ns);
+  const renamedt<ssa_exprt, L2> l2_lhs = set_l2_indices(std::move(lhs), ns);
+  lhs = l2_lhs.get();
```
As for L0/L1, I don't think `l2_lhs` adds value.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -330,7 +326,11 @@ exprt goto_symex_statet::rename(exprt expr, const namespacet &ns)
+        {
+          const renamedt<ssa_exprt, L2> l2_ssa =
+            set_l2_indices(std::move(ssa), ns);
+          ssa = l2_ssa.get();
```
As above.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -455,8 +454,9 @@ bool goto_symex_statet::l2_thread_read_encoding(
+    const renamedt<ssa_exprt, L2> l2_false_case = set_l2_indices(ssa_l1, ns);
 
-    to_ssa_expr(tmp.false_case()).set_level_2(a_s_read.first);
+    if_exprt tmp{cond.as_expr(), l2_true_case.get(), l2_false_case.get()};
```
Maybe we can delay the construction even further and just use `l2_true_case`, `l2_false_case` in the branches below? Overall your changes here indeed make the code much nicer!

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -478,8 +478,9 @@ bool goto_symex_statet::l2_thread_read_encoding(
-    expr=ssa_l1;
+    const renamedt<ssa_exprt, L2> ssa_l2 =
+      set_l2_indices(std::move(ssa_l1), ns);
+    expr = ssa_l2.get();
```
As above: get rid of `ssa_l2`.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -495,15 +496,16 @@ bool goto_symex_statet::l2_thread_read_encoding(
-    expr=ssa_l1;
+    const renamedt<ssa_exprt, L2> ssa_l2 =
+      set_l2_indices(std::move(ssa_l1), ns);
+    expr = ssa_l2.get();
```
No need for `ssa_l2`.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -495,15 +496,16 @@ bool goto_symex_statet::l2_thread_read_encoding(
-  set_l2_indices(ssa_l1, ns);
-  expr=ssa_l1;
+  const renamedt<ssa_exprt, L2> ssa_l2 = set_l2_indices(std::move(ssa_l1), ns);
+  expr = ssa_l2.get();
```
As above.

tautschnig commented on src/goto-symex/renaming_level.cpp:
```diff
@@ -67,6 +67,8 @@ operator()(renamedt<ssa_exprt, L0> l0_expr) const
 operator()(renamedt<ssa_exprt, L1> l1_expr) const
 {
+  if(!l1_expr.get().get_level_2().empty())
+    return renamedt<ssa_exprt, L2>{std::move(l1_expr.value)};
```
This change seems surprising to me, but I'm likely just missing something. Would you mind explaining?

romainbrenguier replied:
As for the symex_level1t case, it does not change the behaviour because it is only called from set_l2_indices which had the same check until this commit. We need the check to be here and not in `set_l2_indices`, because `set_l2_indices` cannot create `renamedt` expression, the constructor is private, this ensures we cannot make a `renamedt` without checking it has the right indices set.

tautschnig reviewed (approved):
Certainly adds clarity, but I think we can reduce the size of the change quite a bit as indicated below. And some parts are still unclear to me.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 1bb05b68eb8d3dbc159d01e5f67edadab2c49a98).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103359871
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.
